### PR TITLE
Reuse CAClient.CreateRequest for NodeAgent server.

### PR DESCRIPTION
**Pull request by incfly**
- Refactoring CAClient to support fetching keycert for
multiple identity.
- KeyCertBundle adds a CertOptions method for CAClient to invoke.

### CODE REVIEW

File: security/pkg/caclient/keycertbundlerotator.go [hunk 1]
```diff
@@ -87,7 +87,12 @@ func (c *keyCertBundleRotatorImpl) Start(errCh chan<- error) {
 			}
 		}
 		log.Infof("Retrieve new key and certs.")
-		certBytes, certChainBytes, privateKeyBytes, err := c.retriever.Retrieve()
+		co, err := c.keycert.CertOptions()
+		if err != nil {
+			errCh <- fmt.Errorf("failed to extact CertOptions from bundle %v", err)
```
Comment (myidpt) [inline security/pkg/caclient/keycertbundlerotator.go]: Before return, you need to set:
c.stoppedMutex.Lock()
c.stopped = true
c.stoppedMutex.Unlock()

File: security/pkg/caclient/client.go [hunk 2]
```diff
@@ -27,51 +27,35 @@ import (
...
 // CAClient is a client to provision key and certificate from the upstream CA via CSR protocol.
 type CAClient struct {
-	platformClient platform.Client
-	protocolClient grpc.CAGrpcClient
-	istioCAAddress string
-
-	identity    string
-	identityOrg string
-	rSAKeySize  int
-	ttl         time.Duration
```
Comment (myidpt) [inline security/pkg/caclient/client.go]: I think we still need these fields... How does node agent get the first key/cert for a new workload, if we don't have these specs?

File: security/cmd/node_agent/management/server.go [hunk 3]
```diff
@@ -39,12 +40,16 @@ import (
...
 	handlerMap map[string]handler.WorkloadHandler
+
 	// makes mgmt-api server to stop
-	done     chan bool
-	config   *na.Config
+	done   chan bool
+	config *na.Config
+	// TODO(incfly): remove these two after they're merged into CAClient.
 	pc       platform.Client
 	cAClient cagrpc.CAGrpcClient
```
[the remark this answers is not in the document]
Comment (incfly) [reply security/cmd/node_agent/management/server.go]: Done

File: security/pkg/caclient/client.go (hunk 2 above)
Comment (incfly) [reply security/pkg/caclient/client.go]: In management/server.go, WorkloadAdded, the identity is retrieved from request.Attrs.ServiceAccount, the Org and RSAKeySize are part of CAClientConfig (a field of NodeAgent server). The value will be passed from cmd flag.

File: security/pkg/caclient/keycertbundlerotator.go (hunk 1 above)
Comment (incfly) [reply security/pkg/caclient/keycertbundlerotator.go]: Thanks! I add a defer to make sure we always do this clean-up when function exits.

File: security/pkg/pki/util/keycertbundle.go [hunk 4]
```diff
@@ -151,6 +154,25 @@ func (b *KeyCertBundleImpl) VerifyAndSetAll(certBytes, privKeyBytes, certChainBy
...
+// CertOptions returns the certificate config based on currently stored cert.
+func (b *KeyCertBundleImpl) CertOptions() (*CertOptions, error) {
+	b.mutex.RLock()
+	defer b.mutex.RUnlock()
+	ids, err := ExtractIDs(b.cert.Extensions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract id %v", err)
+	}
+	if len(ids) != 1 {
+		return nil, fmt.Errorf("expect single id from the cert, found %v", len(ids))
```
Comment (wattli) [inline security/pkg/pki/util/keycertbundle.go]: Print out ids for better debugging purpose?

Comment (incfly) [reply security/pkg/pki/util/keycertbundle.go]: Done.

File: security/pkg/caclient/keycertbundlerotator.go (hunk 1 above)
Comment (myidpt) [reply security/pkg/caclient/keycertbundlerotator.go]: Great. Thanks!

**Comment by wattli** [comment]
please resolve conflict

**Pull request merged**